README.txt
This Plugin is written by Paulsen-IT (http://www.paulsen-it.de)

A live demo you can found under http://www.paulsen-it.de/tutorials/pit_googlemaps.html

The Documentation in English you found in doc/
The Documentation in German you found on the live demo site

ChangeLog
-- 1.0.14	Changes to run in 6.1.x
-- 1.0.13	Changes of API calls
-- 1.0.12	New manual and depends added
-- 1.0.11	Some Bugfixes, and new Google API
-- 1.0.10	Insert Zoom-Level 16 and 17
-- 1.0.9	Upload again because TER doesn't found the Extension anymore
-- 1.0.8	Some JavaScript errors are fixed
-- 1.0.7	You can now disable the Route To ... From in the Info Window 
-- 1.0.6	Change rawurlencode to urlencode 
-- 1.0.5	New Feature: You can choose which map type should be default
-- 1.0.4	Bugfixes
-- 1.0.3	Some Bugfixes, New Feature Sidebar, Manual inserting from GeoData
-- 1.0.2	Bug with caching geo data removed and some changes in manual
-- 1.0.1	some changes in manual
-- 1.0.0	initial release

2009-12-09 Paulsen-IT  <service@paulsen-it.de>

ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) {
	die ('Access denied.');
}

t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY.'_pi1']='layout,select_key';


t3lib_extMgm::addPlugin(array(
	'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.list_type_pi1',
	$_EXTKEY . '_pi1',
	t3lib_extMgm::extRelPath($_EXTKEY) . 'ext_icon.gif'
), 'list_type');

if (TYPO3_MODE == 'BE') {
    $TBE_MODULES_EXT['xMOD_db_new_content_el']['addElClasses']['tx_pitgooglemaps_pi1_wizicon'] = t3lib_extMgm::extPath($_EXTKEY).'pi1/class.tx_pitgooglemaps_pi1_wizicon.php';
}

$tempColumns = array (
	'tx_pitgooglemaps_addresses' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_addresses',		
		'config' => array (
			'type' => 'text',
			'cols' => '30',	
			'rows' => '5',
		)
	),
	'tx_pitgooglemaps_width' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_width',		
		'config' => array (
			'type'     => 'input',
			'size'     => '4',
			'max'      => '4',
			'eval'     => 'int',
			'checkbox' => '0',
			'range'    => array (
				'upper' => '1000',
				'lower' => '10'
			),
			'default' => 0
		)
	),
	'tx_pitgooglemaps_height' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_height',		
		'config' => array (
			'type'     => 'input',
			'size'     => '4',
			'max'      => '4',
			'eval'     => 'int',
			'checkbox' => '0',
			'range'    => array (
				'upper' => '1000',
				'lower' => '10'
			),
			'default' => 0
		)
	),
	'tx_pitgooglemaps_zoom' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom',		
		'config' => array (
			'type' => 'select',
			'items' => array (
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.0', '1'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.1', '2'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.2', '3'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.3', '4'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.4', '5'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.5', '6'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.6', '7'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.7', '8'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.8', '9'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.9', '10'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.10', '11'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.11', '12'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.12', '13'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.13', '14'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.14', '15'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.15', '16'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_zoom.I.16', '17'),
			),
			'size' => 1,	
			'maxitems' => 1,
		)
	),
	'tx_pitgooglemaps_showsidebar' => array (        
        'exclude' => 0,        
        'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showsidebar',        
        'config'  => array (
				'type'    => 'check',
				'default' => '0'
		)
    ),
    'tx_pitgooglemaps_showroute' => array (        
        'exclude' => 0,        
        'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showroute',        
        'config'  => array (
				'type'    => 'check',
				'default' => '1'
		)
    ),
    'tx_pitgooglemaps_showtype' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype',		
		'config' => array (
			'type' => 'select',
			'items' => array (
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.0', '0'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.1', '1'),
				//array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.2', '2'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.3', '3'),
				//array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.4', '4'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.5', '5'),
				array('LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_showtype.I.6', '6'),
			),
			'size' => 1,	
			'maxitems' => 1,
		)
	),
	'tx_pitgooglemaps_infowindow' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_infowindow',		
		'config' => array (
			'type' => 'text',
			'cols' => '30',
			'rows' => '5',
			'wizards' => array(
				'_PADDING' => 2,
				'RTE' => array(
					'notNewRecords' => 1,
					'RTEonly'       => 1,
					'type'          => 'script',
					'title'         => 'Full screen Rich Text Editing|Formatteret redigering i hele vinduet',
					'icon'          => 'wizard_rte2.gif',
					'script'        => 'wizard_rte.php',
				),
			),
		)
	),
	'tx_pitgooglemaps_markers' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_markers',		
		'config' => array (
			'type' => 'text',
			'cols' => '30',	
			'rows' => '5',
		)
	),
	'tx_pitgooglemaps_geodata' => array (		
		'exclude' => 0,		
		'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_geodata',		
		'config' => array (
			'type' => 'text',
			'cols' => '30',	
			'rows' => '5',
		)
	),
	 'tx_pitgooglemaps_markericons' => array (        
        'exclude' => 0,        
        'label' => 'LLL:EXT:pit_googlemaps/locallang_db.xml:tt_content.tx_pitgooglemaps_markericons',        
        'config' => array (
            'type' => 'group',
            'internal_type' => 'file',
            'allowed' => 'gif,png,jpeg,jpg',    
            'max_size' => $GLOBALS['TYPO3_CONF_VARS']['BE']['maxFileSize'],    
            'uploadfolder' => 'uploads/tx_pitgooglemaps',
            'show_thumbs' => 1,    
            'size' => 5,    
            'minitems' => 0,
	    'maxitems' => 100,
        )
    ),
	
);


t3lib_div::loadTCA('tt_content');
t3lib_extMgm::addTCAcolumns('tt_content', $tempColumns, 1);
// t3lib_extMgm::addToAllTCAtypes('tt_content','tx_pitgooglemaps_addresses;;;;1-1-1, tx_pitgooglemaps_width, tx_pitgooglemaps_height, tx_pitgooglemaps_zoom, ' .
//     'tx_pitgooglemaps_infowindow;;;richtext[]:rte_transform[mode=ts], tx_pitgooglemaps_markers, tx_pitgooglemaps_geodata');
$TCA['tt_content']['types']['list']['subtypes_addlist'][$_EXTKEY.'_pi1']='tx_pitgooglemaps_addresses,tx_pitgooglemaps_width, tx_pitgooglemaps_height, tx_pitgooglemaps_zoom, ' .
    'tx_pitgooglemaps_showsidebar, tx_pitgooglemaps_showroute, tx_pitgooglemaps_showtype, tx_pitgooglemaps_infowindow;;;richtext[]:rte_transform[mode=ts], tx_pitgooglemaps_markers, ' .
    'tx_pitgooglemaps_markericons';
?>

pi1/class.tx_pitgooglemaps_pi1.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009 Paulsen-IT <service@paulsen-it.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */

require_once(PATH_tslib.'class.tslib_pibase.php');


/**
 * Plugin 'Google Maps' for the 'pit_googlemaps' extension.
 *
 * @author	Paulsen-IT <service@paulsen-it.de>
 * @package	TYPO3
 * @subpackage	tx_pitgooglemaps
 */
class tx_pitgooglemaps_pi1 extends tslib_pibase {
	var $prefixId      = 'tx_pitgooglemaps_pi1';		// Same as class name
	var $scriptRelPath = 'pi1/class.tx_pitgooglemaps_pi1.php';	// Path to this script relative to the extension dir.
	var $extKey        = 'pit_googlemaps';	// The extension key.
	var $pi_checkCHash = TRUE;

    var $googleMapsApiV3AllowedLanguages = '';
    var $data = array();

	/**
	 * This function checks if the Addresses of the Backend Plugin are changed so we need new geodata
	 * @return unknown_type
	 */
	private function checkSameAddresses() {
		$geodata = unserialize($this->geocache);
		if (is_array($geodata)) {
			$allOk = TRUE;
			if (count($geodata) <> count($this->addresses)) {
				$allOk = FALSE;
            }
			foreach($geodata as $k=>$v) {
				if(trim($this->addresses[$k]) <> $v['cacheAddress']) {
					$allOk = FALSE;
                }
				if($v['geoData']['lon'] == 0 && $v['geoData']['lat'] == 0) {
					$allOk = FALSE;
                }
			}
			if (!$allOk) {
				$this->getNewGeoData();
            }
		} else {
			$this->getNewGeoData();
        }
	}
	
	/**
	 * If the addresses in Backend change we need to get new Geocache Data
	 * @return nothing
	 */
	private function getNewGeoData() {
		foreach($this->addresses as $k=>$v) {
			if (trim($v)<>'') {
				if (is_array($this->userGeoData[$k]) && count($this->userGeoData[$k]) == 3)	{
					$newGeoData[$k]['geoData']['lon'] = $this->userGeoData[$k][1];
					$newGeoData[$k]['geoData']['lat'] = $this->userGeoData[$k][2];
				} else {
					$newGeoData[$k]['cacheAddress'] = trim($v);
					$newGeoData[$k]['geoData'] = $this->geoGetCoords(trim($v));
					if (!is_array($newGeoData[$k]['geoData'])) {
						$newGeoData[$k]['geoData'] = $this->geoGetCoords(trim($v), 'YAHOO');
                    }
					if (!is_array($newGeoData[$k]['geoData'])) {
						$newGeoData[$k]['geoData']['lon'] = 0.00;
						$newGeoData[$k]['geoData']['lat'] = 0.00;
					}
				}
			}
		}
		$this->geocache = serialize($newGeoData);
		$updateArray = array(
            'tx_pitgooglemaps_geodata' => $this->geocache,
            'tstamp' => time(),
        );
		$where = 'uid = '.$this->cObj->data['uid'];
		$query = $GLOBALS['TYPO3_DB']->exec_UPDATEquery('tt_content', $where, $updateArray);
		//$res = $GLOBALS['TYPO3_DB']->sql(TYPO3_db, $query);
		$res = $GLOBALS['TYPO3_DB']->sql_query(TYPO3_db, $query);
	}
	
	/**
	 * get geocode lat/lon points for given address from some geocoding service
	 *
	 * @param string $address
	 * @param string $service
	 */
	private function geoGetCoords($address, $service = 'GOOGLE') {
		// Pre-set result
		$_result = FALSE;
		// Do work according to chosen lookup service
		switch ($service) {
			// Yahoo Geocoding API
			case 'YAHOO':
				// Compose URL
				$_url = 'http://api.local.yahoo.com/MapsService/V1/geocode';
				$_url .= '?appid=' . 'MyMapApp';//$this->app_id;
				$_url .= '&location=' . urlencode($address); //rawurlencode($address);
				$curlHandle = curl_init(); // init curl

				// options
				curl_setopt($curlHandle, CURLOPT_URL, $_url); // set the url to fetch
				curl_setopt($curlHandle, CURLOPT_HEADER, 0); // set headers (0 = no headers in result)
				curl_setopt($curlHandle, CURLOPT_RETURNTRANSFER, 1); // type of transfer (1 = to string)
				curl_setopt($curlHandle, CURLOPT_TIMEOUT, 10); // time to wait in seconds
				curl_setopt($curlHandle, CURLOPT_CONNECTTIMEOUT, 5);

				$_result = curl_exec($curlHandle); // make the call

				curl_close($curlHandle); // close the connection
				if($_result<>'') {
				//if ($_result = file_get_contents($_url)) {
					if (preg_match('/<Latitude>(-?\d+\.\d+)<\/Latitude><Longitude>(-?\d+\.\d+)<\/Longitude>/', $_result, $_match)) {
						$_coords['lon'] = $_match[2];
						$_coords['lat'] = $_match[1];
						$_result = TRUE;
					}
				}
				break;
			// Google Maps API
			case 'GOOGLE':
			default:
				// Compose URL
				$_url = 'http://maps.google.com/maps/api/geocode/json?address=' . urlencode($address) . '&sensor=false';
				$curlHandle = curl_init(); // init curl

				// options
				curl_setopt($curlHandle, CURLOPT_URL, $_url); // set the url to fetch
				curl_setopt($curlHandle, CURLOPT_HEADER, 0); // set headers (0 = no headers in result)
				curl_setopt($curlHandle, CURLOPT_RETURNTRANSFER, 1); // type of transfer (1 = to string)
				curl_setopt($curlHandle, CURLOPT_TIMEOUT, 10); // time to wait in seconds
				curl_setopt($curlHandle, CURLOPT_CONNECTTIMEOUT, 5);

				$_result = curl_exec($curlHandle); // make the call

				curl_close($curlHandle); // close the connection
				if($_result<>'') {
					$resp = json_decode($_result, TRUE);
					if($resp['status']='OK') {
						$_coords['lon'] = $resp['results'][0]['geometry']['location']['lng'];
						$_coords['lat'] = $resp['results'][0]['geometry']['location']['lat'];
						$_result = TRUE;
					}
				}
			}
		
		// Return coordinates or FALSE
		if ($_result) {
			return $_coords;
		}
		return $_result; 
	}
	
	/**
	 * The center position of Google Maps is needed, here we calculate it
	 * @return array
	 */
	private function gettingCenterOfMap() {
		$geodata = unserialize($this->geocache);
		if (is_array($geodata) && count($geodata) > 1) {
			foreach($geodata as $gd) {
				$lons[] = $gd['geoData']['lon'];
				$lats[] = $gd['geoData']['lat'];
			}
			sort($lons, SORT_NUMERIC);
			sort($lats, SORT_NUMERIC);
		
			$biggestlon = array_pop($lons);
			$biggestlat = array_pop($lats);
			
			$smallestlon = array_shift($lons);
			$smallestlat = array_shift($lats);
		
			return array(
                'lon' => round(($biggestlon+$smallestlon)/2, 3),
                'lat' => round(($biggestlat+$smallestlat)/2, 3)
            );
		} elseif (is_array($geodata) && count($geodata) == 1) {
			return array(
                'lon' => $geodata[0]['geoData']['lon'],
                'lat' => $geodata[0]['geoData']['lat']
            );
		}
	}
	
	/**
	 * Google needs some JavaScript to generate the Point (Markers) on Maps, here the JS-Code will be generated
	 * @return string
	 */
	private function gettingJSforPointers() {
		$geodata = unserialize($this->geocache);
		$js = '';
		$jsIcon = '';
        $currentIcon = '';
        $lastIcon = '';
		asort($this->addresses);
		$id = -1;
		if ($this->markerImages[0] <> '') {
			$lastIcon = $this->markerImages[0];
        }
		foreach($this->addresses as $k => $v) {
			$id++;
			$gd = $geodata[$k];
			$htmlInfo = '';
			if (is_array($this->marker[$k])) {
				$htmlInfo = $this->infowindow;
				foreach($this->marker[$k] as $key=>$value) {
					$htmlInfo = str_replace($key, $value, $htmlInfo);
                }
				$htmlInfo = str_replace(array("\r", "\n", "\r\n", "\n\r"), '', $htmlInfo);
				$htmlInfo = str_replace("'", "\\'", $htmlInfo);
			}
            // mb, 2013-12-08: This logic looks strange. Wrong?
			// The Icon would be created for Google, if some exists
			if (is_array($this->markerImages)) {
				//if($this->markerImages[$k] <> '')
				$currentIcon = $this->markerImages[$k];
            }
                if($currentIcon == '') {
                    $currentIcon = $lastIcon;
                }
                if($currentIcon <> '') {
                    $icon = 'icon: "' . $this->upload . $currentIcon . '",';
                }

			// The point for google is created
			$js .= '
							markerTitle_' . $this->uid . '[' . $id . '] = "' . preg_replace("/\r|\n/s", '', $v) . '";
							point_' . $this->uid . '[' . $id . '] = new google.maps.LatLng(' . $gd['geoData']['lat'] . ',' . $gd['geoData']['lon'] . ');
							marker_' . $this->uid . '_' . $id . '_Options = {
								map: map,
								' . $icon . '
								position: point_' . $this->uid . '[' . $id . '],
								title: "' . preg_replace("/\r|\n/s", '', $v) . '",
							}
							marker_' . $this->uid . '[' . $id . '] = new google.maps.Marker(marker_' . $this->uid . '_' . $id . '_Options);

			';
			if ($this->cObj->data['tx_pitgooglemaps_showroute']) {
				$js .= '
								marker_' . $this->uid . '_' . $id . '_InfoWindowHtml = new google.maps.InfoWindow({ content:\'' . $htmlInfo .
                                    ' <p class="tx-pitgooglemaps-pi1_route">' . htmlspecialchars($this->data['route']) . '<a href="http://maps.google.com/maps?daddr=\'+point_' . $this->uid .
                                    '[' . $id . ']+\'" target="_blank">' . htmlspecialchars($this->data['toHere']) . '</a> - <a href="http://maps.google.com/maps?saddr=\'+point_' . $this->uid .
                                    '[' . $id . ']+\'" target="_blank">' . htmlspecialchars($this->data['fromHere']) . '</a></p>\'});
				';
            } else {
				$js .= '
								marker_' . $this->uid . '_' . $id . '_InfoWindowHtml = new google.maps.InfoWindow({ content: \'' . $htmlInfo . '\'});
				';
            }
			$js .= '
							google.maps.event.addListener(marker_' . $this->uid . '[' . $id . '], "click", function() {
							    marker_' . $this->uid . '_' . $id . '_InfoWindowHtml.open(map,marker_' . $this->uid . '[' . $id . ']);
							});
						
			';
		}
		$js = $jsIcon . $js;
		return $js;
	}
		
	function createSidebar() {
		$sidebarContent = '
			function createSidebar' . $this->uid . '() {
				var html = "";
				for (var i = 0; i < marker_' . $this->uid . '.length; i++) {
					html += \'<a id="marker\'+i+\'" href="javascript:sidebar_click' . $this->uid .
                        '(\'+i+\')" onmouseover="javascript:markerhover_in' . $this->uid .
                        '(\'+i+\')" onmouseout="javascript:markerhover_out' . $this->uid .
                        '(\'+i+\')">\' + markerTitle_' . $this->uid . '[i] + \'</a><br />\';
				}
				document.getElementById("sidebar' . $this->uid . '").innerHTML = html;
			}
			function sidebar_click' . $this->uid . '(id) {
				window.setTimeout("google.maps.event.trigger(marker_' . $this->uid . '["+id+"],\'click\')", 500);
			}
			function markerhover_in' . $this->uid . '(id) {
				google.maps.event.trigger(marker_' . $this->uid . '[id],\'click\');
			}
			function markerhover_out' . $this->uid . '(id) {
				setTimeout("marker_' . $this->uid . '_"+id+"_InfoWindowHtml.close();",1)
			}
		';
		return $sidebarContent;
	}
		
	/**
	 * The main method of the PlugIn
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	function main($content, $conf) {
		$this->conf = $conf;
		$this->pi_setPiVarDefaults();
		$this->pi_loadLL();
		$this->upload = 'uploads/tx_pitgooglemaps/';
        $this->googleMapsApiV3AllowedLanguages = ',ar,eu,bg,bn,ca,cs,da,de,el,en,en-AU,en-GB,es,eu,fa,fi,fil,fr,' .
            'gl,gu,hi,hr,hu,id,it,iw,ja,kn,ko,lt,lv,ml,mr,nl,nn,no,pl,' .
            'pt,pt-BR,pt-PT,ro,ru,sk,sl,sr,sv,tl,ta,te,th,tr,uk,vi,zh-CN,zh-TW,';

        $this->data = array();
        $this->data['route'] = $this->pi_getLL('route', 'Route:');
        if (strlen($this->data['route'])) {
            $this->data['route'] .= ' ';
        }
        $this->data['fromHere'] = $this->pi_getLL('fromHere', 'From here');
        $this->data['toHere'] = $this->pi_getLL('toHere', 'To here');

        $keys = array();
        $temp = array(
            'DE_DE' => 'de',
            'AT_DE' => 'de',
            'BE_FR' => 'fr',
            'BE_NL' => 'nl',
            'CH_DE' => 'de',
            'CH_FR' => 'fr',
            'CN_CN' => 'zh-CN',
            'FR_FR' => 'fr',
            'IN_EN' => 'en',
            'LU_DE' => 'de',
            'LU_FR' => 'fr',
            'NL_NL' => 'nl',
            'RU_RU' => 'ru',
            'UK_EN' => 'en-GB',
            'PT_PT' => 'pt',
            'PL_PL' => 'pl',
        );
        $k = $temp[$this->LLkey];
        if (strlen($k)) {
            $keys[] = $k;
        } else {
            $keys[] = $this->LLkey; // config.language
        }
        if (0) { // let's disable this for now
            $p = strpos($this->LLkey, '_'); // something like 'de_DE'?
            if (is_int($p) && $p > 0) {
                $keys[] = strtolower(substr($this->LLkey, 0, $p));
            }
        }
        // this parameter is used to localize the map.
        // https://www.google.de/search?q=google+maps+api+v3+localization
        // https://developers.google.com/maps/documentation/javascript/examples/map-language
        $this->data['googleMapsLanguageParam'] = '';
        foreach ($keys as $k) {
            if (stripos($this->googleMapsApiV3AllowedLanguages, ',' . $k . ',') !== FALSE) {
                $this->data['googleMapsLanguageParam'] = '&language=' . $k;
            }
        }

		// Set Google Maps Java-Files to Header
		$GLOBALS['TSFE']->additionalHeaderData[] = '<script type="text/javascript" src="https://maps.googleapis.com/maps/api/js?sensor=false' . $this->data['googleMapsLanguageParam'] . '"></script>';
		
		// Make a Div for the Map minimum Height = 100, minimum width = 100
		$width = (int) $this->cObj->data['tx_pitgooglemaps_width'];
		if ($width < 100) {
			$width = 100;
        }
		$height = (int) $this->cObj->data['tx_pitgooglemaps_height'];
		if ($height < 100) {
			$height = 100;
        }
		$this->uid = $this->cObj->data['uid'];
		if ($this->cObj->data['tx_pitgooglemaps_showsidebar']) {
			$mapclass = 'pit_googlemaps-map-withsidebar';
        } else {
			$mapclass = 'pit_googlemaps-map';
        }
		$content = '<div class="pit_googlemaps-wrap clearfix">
						<div class="' . $mapclass . '" id="map_canvas' . $this->uid .
                        '" style="width:' . $width . 'px; height:' . $height . 'px;"></div>';
		
		if($this->cObj->data['tx_pitgooglemaps_showsidebar']) {
			$content .= '
						<div class="pit_googlemaps-sidebar" id="sidebar' . $this->uid . '"></div>';
        }
		$content .= '
					</div>
		';
		
		// Splitting Informations
		$this->addresses = explode("\n", $this->cObj->data['tx_pitgooglemaps_addresses']);
		
		// Clean Adresses from Breaks and empty Lines
		foreach($this->addresses as $addr) {
			$cleanaddr = preg_replace("/\r|\n/s", '', $addr);
			if ($cleanaddr <> '') {
				$newAddr[] = $cleanaddr;
            }
		}
		$this->addresses = $newAddr;
		$this->geocache = $this->cObj->data['tx_pitgooglemaps_geodata'];

		// If User have set geodata manual
		foreach($this->addresses as $k=>$add) {
			$userGeo = strstr($add, '|');
			if (!($userGeo===FALSE)) {
				$this->userGeoData[$k] = explode('|', $userGeo);
				$this->addresses[$k] = substr($add, 0, strpos($add, '|'));
			}
		}
		//var_dump($this->userGeoData);
		$markers = explode("\n", $this->cObj->data['tx_pitgooglemaps_markers']);
		$this->infowindow = $this->cObj->data['tx_pitgooglemaps_infowindow'];
		$this->markerImages = explode(',', $this->cObj->data['tx_pitgooglemaps_markericons']);
		
		// Getting markers in right form
		$m = array();
		foreach($markers as $x) {
			array_push($m, explode('|', $x));
        }
		foreach($m as $k=>$v) {
			foreach($v as $x) {
				$z = explode('=', $x);
				$this->marker[$k][$z[0]] = trim($z[1]);
			}
        }
		
		// Checking if geocache is right one to addresses
		$this->checkSameAddresses();
		
		// Setting the center of the map
		$mapCenter = $this->gettingCenterOfMap();
		
		// Getting JavaScript-Code of all Points
		$pointerJS = $this->gettingJSforPointers();
		
		// Create Content for the Sidebar 
		if ($this->cObj->data['tx_pitgooglemaps_showsidebar']) {
			$sidebarContent = $this->createSidebar();
        }
		
		$map_type[0] = 'google.maps.MapTypeId.ROADMAP';
		$map_type[1] = 'google.maps.MapTypeId.SATELLITE';
		$map_type[3] = 'google.maps.MapTypeId.HYBRID';
		$map_type[5] = 'google.maps.MapTypeId.TERRAIN';
		$map_type[6] = 'google.maps.MapTypeId.SATELLITE';
		
		if ($this->cObj->data['tx_pitgooglemaps_showtype'] >= 0) {
			$map_typeJS = 'map.setMapTypeId(' . $map_type[$this->cObj->data['tx_pitgooglemaps_showtype']] . ');';
        }
		
		$content .= '
					<script type="text/javascript">
					map = null;
					var markerIcon_' . $this->uid . ' = new Array();
					var point_' . $this->uid . ' = new Array();
					var marker_' . $this->uid . ' = new Array();
					var markerTitle_' . $this->uid . ' = new Array();
					
					' . $sidebarContent . '
					
					function setupMap()
						{
							var mapOptions = {
								center: new google.maps.LatLng(' . $mapCenter['lat'] . ', ' . $mapCenter['lon'] . '),
								zoom: ' . $this->cObj->data['tx_pitgooglemaps_zoom'] . ',
								};

							map = new google.maps.Map(document.getElementById("map_canvas' . $this->cObj->data['uid'] . '"),mapOptions);
							map.setZoom(' . $this->cObj->data['tx_pitgooglemaps_zoom'] . ');
							' . $map_typeJS . '
        					' . $pointerJS;
		if ($this->cObj->data['tx_pitgooglemaps_showsidebar']) {
			$content .= 	'
        					createSidebar' . $this->uid . '();
        					';
        }
		$content .=	'
							}
					google.maps.event.addDomListener(window, "load", setupMap);
					</script>
		';
	
		return $this->pi_wrapInBaseClass($content);
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pit_googlemaps/pi1/class.tx_pitgooglemaps_pi1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pit_googlemaps/pi1/class.tx_pitgooglemaps_pi1.php']);
}

?>

ext_emconf.php
<?php

/* ######################################################################
 * Extension Manager/Repository config file for ext "pit_googlemaps".
 *
 * Auto generated 06-12-2013 20:02
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 * ######################################################################
 */

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Google Maps',
	'description' => 'Show Markers on Google Maps. You have the possibility to put more than one marker into the map and all points can display an info-window, different text can shown (Template based in Backend)',
	'category' => 'plugin',
	'shy' => 0,
	'version' => '1.0.14',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'excludeFromUpdates', //'stable',
	'uploadfolder' => 1,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 1,
	'lockType' => '',
	'author' => 'Paulsen-IT',
	'author_email' => 'service@paulsen-it.de',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'cms' => '',
			'typo3' => '4.0.0-6.1.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:18:{s:9:"ChangeLog";s:4:"a80c";s:10:"README.txt";s:4:"5c10";s:12:"ext_icon.gif";s:4:"fdea";s:12:"ext_icon.xcf";s:4:"82f7";s:17:"ext_localconf.php";s:4:"1d05";s:14:"ext_tables.php";s:4:"a053";s:14:"ext_tables.sql";s:4:"c4db";s:28:"ext_typoscript_constants.txt";s:4:"d623";s:24:"ext_typoscript_setup.txt";s:4:"8a6d";s:13:"locallang.xml";s:4:"839a";s:16:"locallang_db.xml";s:4:"9aba";s:14:"doc/manual.sxw";s:4:"6dea";s:19:"doc/wizard_form.dat";s:4:"ef55";s:20:"doc/wizard_form.html";s:4:"c7a7";s:14:"pi1/ce_wiz.gif";s:4:"b092";s:34:"pi1/class.tx_pitgooglemaps_pi1.php";s:4:"1f39";s:42:"pi1/class.tx_pitgooglemaps_pi1_wizicon.php";s:4:"8939";s:17:"pi1/locallang.xml";s:4:"ab53";}',
);

?>